doc/man/man1/reihe.1
\"    This documentation is copyright 1994 David Huron.
.TH reihe 1 "1994 Dec. 4"
.AT 3
.sp 2
.SH "NAME"
.in +2
.in +11
.ti -11
\fBreihe\fR  --  output specified row variant for a given prime row
.in -11
.in -2
.sp 1
.sp 1
.SH "SYNOPSIS"
.in +2
\fBreihe  -\fR[\fBI\fR|\fBi\fR]  \fIn\fR  [\fB-a\fR]  [\fB-m  \fIn\fR]  \fIprimefile\fR
.br
\fBreihe  -\fR[\fBP\fR|\fBp\fR]  \fIn\fR  [\fB-a\fR]  [\fB-m  \fIn\fR]  \fIprimefile\fR
.br
\fBreihe  -\fR[\fBR\fR|\fBr\fR]  \fIn\fR  [\fB-a\fR]  [\fB-m  \fIn\fR]  \fIprimefile\fR
.br
\fBreihe  -\fR[\fBR\fR|\fBr\fR]  \fIprimefile\fR
.br
\fBreihe  -\fR[\fBRI\fR|\fBri\fR]  \fIn\fR  [\fB-a\fR]  [\fB-m  \fIn\fR]  \fIprimefile\fR
.br
\fBreihe  -\fR[\fBS\fR|\fBs\fR]  \fIn\fB  \fIprimefile\fR
.in -2
.sp 1
.sp 1
.SH "DESCRIPTION"
.in +2
The
.B "reihe"
command outputs a specified row variant for a given prime row input.
Normally,
.B "reihe"
is used to generate tone-row variants for 12-tone rows.
However, the \(odtone-rows\(cd can consist of any number of pitch-classes
and need not be based on modulo-12 class-equivalence.
In certain circumstances the \(odprime row\(cd may consist of non-numeric
or non-pitch-related data -- such as articulation marks or dynamic marks.
In addition to the traditional set-theoretic transformations,
.B "reihe"
also permits cyclic rotation (shifting) of set elements.
.sp 1
.sp 1
The input to the
.B "reihe"
command is interpreted as a
.I "prime"
or basic set of elements from which a transformed set
is to be derived and output.
Four traditional types of set-transformations can be generated by
specifying the appropriate option:
.B "-P"
(prime),
.B "-I"
(inversion),
.B "-R"
(retrograde),
and
.B "-RI"
(retrograde-inversion).
Either upper- or lower-case characters can be used when specifying
these options.
.sp 1
.sp 1
Following each option is a numerical parameter.
For prime and inversion forms, the numerical parameter indicates
the pitch-class value for the first \(odnote\(cd of the output form.
For example,
.B "-P 0"
specifies the prime form beginning on pitch-class 0;
.B "-P 1"
specifies the prime transposed so that it begins on pitch-class 1.
For retrograde and retrograde-inversion forms, the numerical parameter
specifies the
.I "last"
value of the output row.
.sp 1
.sp 1
Positive integers greater than 11 are permitted in the input --
but are treated as modulo-12 equivalent (unless the
.B "-m"
option specifies a different modulus).
In the case of \f(CR**pc\fR inputs, the upper-case letters A and B
are accepted as aliases for pitch-classes 10 and 11 respectively.
Following the Fortean practice, pitch-classes 10 and 11 may
alternatively be represented in the input by the upper-case
letters T (ten) and E (eleven) -- although this latter convention
is discouraged.
.sp 1
.sp 1
In addition to the traditional set-transformations of transposition,
inversion, retrograde, and retrograde-inversion,
.B "reihe"
also provides for rotation -- where the set elements
are cyclically
.I "shifted"
by a specified number of positions.
For example,
.B "-s 3"
causes each set element to be shifted forward by 3 positions.
The shift transformation can be combined with each of the other
traditional transformations only by invoking the
.B "reihe"
command twice in succession.
.sp 1
.sp 1
Normally,
.B "reihe"
is used to transform numerical data (typically pitch-class values).
However, the retrograde (\fB-r\fR) and shift (\fB-s\fR) operations
can be performed on any data (including non-numeric data
-- such as articulation marks).
For non-numeric data, the retrograde option must be invoked without
a numerical parameter.
Attempting to transpose non-numeric data will result in an error.
.sp 1
.sp 1
The prime-form input file must contain a single spine.
For untransposed retrogrades and rotational shift transformations,
any single-spine Humdrum input will be accepted.
If the
.B "-m"
option is invoked, any input interpretation will be accepted provided
all data tokens contain numbers only.
In the case of tranposed prime, inversion, and retrograde transformations,
the input must conform to the Humdrum pitch-class (\f(CR**pc\fR) representation.
In all cases, comments, null tokens, and tandem interpretations in the
input spine are ignored and are not echoed in the output.
In the case of \f(CR**pc\fR inputs, barlines and rests are also ignored.
Output interpretations always echo the input interpretation.
.sp 1
.sp 1
By way of example, consider the following input file \f(CRwebern\fR:
.in +2
.sp 1
.TS
l.
!! Anton von Webern
!! Klavierstueck, opus posthumous
**pc
\(eq1
9
10
\.
11
8
7
\(eq2
1
2
3
r
6
5
4
\(eq3
0
*-
.TE
.sp 1
.in -2
Executing the command:
.in +2
reihe -p 1 webern
.in -2
produces the following output:
.in +2
.sp 1
.TS
l.
**pc
1
2
3
0
11
5
6
7
10
9
8
4
*-
.TE
.sp 1
.in -2
Notice that the comments, barlines, rests, and null tokens have
been eliminated from the input file.
This leaves the output in a form better suited to pattern matching
using the
.B "patt"
or
.B "pattern"
commands.
.sp 1
.sp 1
Similarly, executing the command:
.in +2
reihe -r 1 webern
.in -2
produces:
.in +2
.sp 1
.TS
l.
**pc
1
5
6
7
4
3
2
8
9
0
11
10
*-
.TE
.sp 1
.in -2
.in -2
.SH "OPTIONS"
.in +2
The
.B "reihe"
command provides the following options:
.sp 1
.TS
l l.
\fB-a\fR	for \f(CR**pc\fR inputs, output alphanumeric representation
	  (where A=10, B=11)
\fB-h\fR	displays a help screen summarizing the command syntax
\fB-I \fIn\fR	output inversion set-form starting on pitch-class \fIn\fR
\fB-i \fIn\fR	same as \fB-I\fR option
\fB-m \fIn\fR	calculate according to modulo \fIn\fR arithmetic
\fB-P \fIn\fR	output prime set-form starting on pitch-class \fIn\fR
\fB-p \fIn\fR	same as \fB-P\fR option
\fB-R\fR	output retrograde of input row
\fB-R \fIn\fR	output retrograde set-form ending on pitch-class \fIn\fR
\fB-r\fR	same as \fB-R\fR option
\fB-r \fIn\fR	same as \fB-R \fIn\fR option
\fB-RI \fIn\fR	output retrograde-inversion set-form ending on pitch-
	  class \fIn\fR
\fB-ri \fIn\fR	same as \fB-RI\fR option
\fB-S \fR[\fI\(+-\fR]\fIn\fR	output set-form shifted \fIn\fR elements forward (+)
	  or backward (-)
\fB-s \fR[\fI\(+-\fR]\fIn\fB	same as \fB-S\fR option
.TE
.sp 1
Options are specified in the command line.
.sp 1
.sp 1
When the
.B "-a"
option is invoked, pitch-class inputs (\f(CR**pc\fR) will produce
pitch-class outputs using the alias values `A' for pitch-class 10,
and `B' for pitch-class 11.
(See the \fB**pc\fR representation.)
.sp 1
.sp 1
By default, the
.B "reihe"
command assumes modulo 12 arithmetic for prime, inversion,
retrograde, and retrograde-inversion transformations.
In other words, transposing the numerical value `11' up three pitch-classes
results in an output value of `2.'
The
.B "-m"
option can be used to specify some other modulo value.
If this option is invoked with \f(CR**pc\fR input,
the alias values (A=T=10; B=E=11) are disabled and only
numerical data can be processed and output.
The
.B "-m"
and
.B "-a"
options are thus mutually exclusive.
.in -2
.sp 1
.sp 1
.SH "EXAMPLES"
.in +2
The sample document given below shows a 5-tone row
used in Igor Stravinsky's \(odDirge-Canons\(cd from
.I "In Memoriam Dylan Thomas."
.in +2
.sp 1
.TS
l.
!! I. Stravinsky, 5-tone row
**pc
2
3
6
5
4
*-
.TE
.sp 1
.in -2
The command: \f(CRreihe -s -1 igor\fR   would result in the following
output:
.in +2
.sp 1
.TS
l.
**pc
3
6
5
4
2
*-
.TE
.sp 1
.in -2
The command: \f(CRreihe -a -i 2 igor\fR   would result in the following
output:
.in +2
.sp 1
.TS
l.
**pc
2
1
A
B
0
*-
.TE
.sp 1
.in -2
The command: \f(CRreihe -ri 0 -m 7 igor\fR   would result in the following
output:
.in +2
.sp 1
.TS
l.
**pc
5
4
3
6
0
*-
.TE
.sp 1
.in -2
.in -2
.SH "PORTABILITY"
.in +2
\s-1DOS\s+1 2.0 and up, with the \s-1MKS\s+1 Toolkit.
\s-1OS/2\s+1 with the \s-1MKS\s+1 Toolkit.
\s-1UNIX\s+1 systems supporting the
.I "Korn"
shell or
.I "Bourne"
shell command interpreters, and revised
.I "awk"
(1985).
.in -2
.sp 1
.sp 1
.SH "SEE ALSO"
.in +2
\fB**iv\fR (2), \fBiv\fR (1),
\fB**nf\fR (2), \fBnf\fR (1),
\fBpatt\fR (1), \fBpattern\fR (1),
\fB**pc\fR (2), \fBpc\fR (1),
\fB**pcset\fR (2), \fBpcset\fR (1),
\fB**pf\fR (2), \fBpf\fR (1),
\fBrecode\fR (1),
\fBsemits\fR (1)
.in -2


toolkit/awk/reihe.awk
#####################################################################
##                          REIHE.AWK
# Programmed by:  David Huron     Date: June, 1994
#
# Modifications:
#   Date:      Programmer:    Description:
# June 20/94	Tim Racinsky	Added parse_command function
# July 12/94	Tim Racinsky	Fixed error with -r option when no integer is
#						given (used to change A's, B's, etc. to integers)
#
# The purpose of this program is to output a specified set-complex
# variant for a given prime-form tone row.
#
# Main Functions used:
#	parse_command()		transpose()			invert()
#	retrograde()			shift()
#
#	VARIABLES:
#
#  -stderr:  variable holding destination of error messages
#  -first_file:  flag to indicate if this is the first file to be processed
#  -no_interpretation_yet:  flag to indicate if first interp. record processed
#  -options:  Holds options specified by the user
#  -setform: Holds the option indicating the type of row variant
#            operation: [Pp]-prime, [Ii]-inversion, [Rr]-retrograde,
#            [X]-retrograde-inversion, [Ss]-cyclic rotation (shift)
#  -integer: Hold specific integer value of setform, e.g. P1, R8, RI7, etc.
#  -data_count:  Holds number of data records stored
#  -base:  Holds value for modulo arithmetic
#  -interpretation: Holds type of interpretation processed
#  -error: Flag to indicate if error has occurred in the main loop
#
#	ARRAYS:
#
#  -current_interp:  Holds exclusive interpretations for currently active spines
#  -pattern:  Holds the pattern to match (read in from template file)
#  -token:  Holds the latest set of records to test against pattern array
#
BEGIN {
	FS = OFS = "\t"
	TRUE = 1 ; FALSE = 0
	#
	# Set standard error to the appropriate 'file'
	#
	#if (ARGV[0] == "nawk") stderr = "/dev/tty"
	#else stderr = "con"
	"echo $CON" | getline stderr
	close("echo $CON")
	#
	pc_number = "[0-9]+|[AB]|[TE]"
	shift_number = "^[+-]?[0-9]+$"
	trans_number = "^(([0-9]+)|([ABTE]))$"
	no_interpretation_yet = TRUE
	error = FALSE
	data_count = 0
	integer = 0
	base = 12
	setform = ""
	options = ""
	parse_command()
}
{
if ($0 !~ /^!!/ && NF != 1)
	{
	print "reihe: ERROR: Only one spine of input is permissible." > stderr
	error = TRUE
	exit
	}
#
# Skip those lines which should not be considered
#
if ($0 == "*-") exit
else if ($0 ~ /^!/) next
else if ($0 ~ /^\*[^*]/) next
else if ($0 ~ /[=]/) next
else if ($0 ~ /r/ && integer != "null" && setform !~ /[Ss]/) next
else if ($0 ~ /^\.$/) next
else if ($0 ~ /^\*\*/)
	{
	#
	# Only one exclusive interpretation is allowed per file
	#
	if (no_interpretation_yet)
		{
		if (integer != "null" && setform !~ /[Ss]/ && $0 != "**pc")
			{
			print "reihe: ERROR: Input must be of type **pc." > stderr
			error = TRUE
			exit
			}
		else interpretation = $0
		no_interpretation_yet = FALSE
		}
	else
		{
		print "reihe: ERROR: Cannot specify new exclusive interpretation"\
			 "." > stderr
		error = TRUE
		exit
		}
	}
else
	{
	#
	# Store integer values if processing pc input
	#
	if (integer != "null" && setform !~ /[Ss]/)
		{
		if (match($0,pc_number) > 0)
			{
			token = substr($0,RSTART,RLENGTH)
			prime[++data_count] = token
			}
		}
	#
	# Otherwise, store all data values
	#
	else prime[++data_count] = $0
	}
}
END	{
	#
	# Quit the program if an error occurred or there were no data values
	#
	if (error || data_count == 0) exit
	#
	# Change any A's or B's to 10's and 11's if processing pc input
	#
	if (integer != "null" && setform !~ /[Ss]/)
		{
		for (i = 1; i <= data_count; i += 1)
			{
			if (prime[i] == "A" || prime[i] == "T") prime[i] = 10
			else if (prime[i] == "B" || prime[i] == "E") prime[i] = 11
			}
		#
		# Change all input values to modulo the base
		#
		if (integer != "null" && setform !~ /[Ss]/)
			{
			for (i = 1; i <= data_count; i += 1)
				prime[i] = ((prime[i] % base) + base) % base
			}
		}
	#
	# Process primes.
	#
	if (setform ~ /[pP]/) transpose(prime,final)
	#
	# Process retrogrades.
	#
	else if (setform ~ /[rR]/)
		{
		#
		# If processing **pc input, perform the transpose as well
		#
		if (integer != "null")
			{
			transpose(prime,mid)
			retrograde(mid,final)
			}
		else retrograde(prime,final)
		}
	#
	# Process inversions.
	#
	else if (setform ~ /[Ii]/)
		{
		invert(prime,mid)
		transpose(mid,final)
		}
	#
	# Process retrograde-inversions.
	#
	else if (setform ~ /X/)
		{
		invert(prime,mid1)
		transpose(mid1,mid2)
		retrograde(mid2,final)
		}
	#
	# Process cyclic rotations.
	#
	else if (setform ~ /[Ss]/) shift(prime,final)
	#
	# Change any 10's and 11's to A's and B's if -a option selected
	#
	if (options ~ /a/)
		{
		for (i = 1; i <= data_count; i += 1)
			{
			if (final[i] == 10) final[i] = "A"
			else if (final[i] == 11) final[i] = "B"
			}
		}
	#
	# Output the final array
	#
	print interpretation
	for (i = 1; i <= data_count; i += 1) print final[i]
	print "*-"
	}

############################################################################
#			FUNCTIONS USED IN REIHE.AWK
############################################################################

###########################################################################
##				Function parse_command
#	This function ensures the correctness of the command line options
#
function parse_command()
	{
	setform = ARGV[2]
	#
	# Check to see if -a and -m options were both specified
	#
	if (ARGV[3] != "null")
		{
		if (ARGV[3] ~ /a/ && ARGV[3] ~ /m/)
			{ 
			print "reihe: ERROR: Cannot specify both -a and -m options"\
				 "." > stderr
			error = TRUE
			exit
			}
		else options = ARGV[3]
		}
	#
	# Make sure the value of 'integer' is correct depending on the options
	#
	if (setform ~ /[Ss]/)
		{
		if (ARGV[4] !~ shift_number)
			{
			print "reihe: ERROR: Invalid shift specified: " ARGV[4] > stderr
			error = TRUE
			exit
			}
		else integer = ARGV[4]
		}
	else if (ARGV[4] != "null" && ARGV[4] !~ trans_number)
		{
		print "reihe: ERROR: Invalid integer specified: " ARGV[4] > stderr
		error = TRUE
		exit
		}
	else
		{
		if (ARGV[4] == "A" || ARGV[4] == "T") integer = 10
		else if (ARGV[4] == "B" || ARGV[4] == "E") integer = 11
		else integer = ARGV[4]
		}
	if (integer == "null")
		{
		if (options ~ /[am]/)
			{ 
			print "reihe: ERROR: Cannot specify -a or -m options with "\
				 "-R option\n              when processing non-pc"\
				 "input." > stderr
			error = TRUE
			exit
			}
		}
	if (ARGV[5] != "null") base = ARGV[5]
	ARGV[1] = ARGV[2] = ARGV[3] = ARGV[4] = ARGV[5] = ""
	}

###########################################################################
##				Function transpose
#	This function transposes the values in array input according to integer
#
function transpose(input,output   ,offset,i)
	{
	offset = (((input[1] - integer) % base) + base) % base
	for (i = 1; i <= data_count; i += 1)
		output[i] = (((input[i] - offset) % base) + base) % base
	}

###########################################################################
##				Function retrograde
#	This function reverses the order of the array input
#
function retrograde(input,output   ,i)
	{
	for (i = 1; i <= data_count; i += 1) output[data_count+1-i] = input[i]
	}

###########################################################################
##				Function invert
#	This function inverts the elements of the array input
#
function invert(input,output   ,i)
	{
	for (i = 1; i <= data_count; i += 1)
		output[i] = (((input[1] - input[i] + input[1]) % base) + base) % base
	}

###########################################################################
##				Function shift
#	This function shifts the elements in input 'integer' positions
#
function shift(input,output   ,i)
	{
	for (i = 1; i <= data_count; i += 1)
		output[i] \
= input[(((((i - integer - 1) % data_count) + data_count) % data_count) + 1)]
	}
